magazine.c

```diff
@@ -23,123 +25,67 @@ struct probe_window {
-  const struct probe_window *current_event = &mag_state.probe_events[mag_state.evt_no];
-  const struct probe_window *last_event = &mag_state.probe_events[last_evt_no];
-
-  if (mag_state.mag_events < 2) {
-    return false;
-  }
-
-  return current_event->direction != last_event->direction;
-}
+  while (!queue_is_empty(&edge_events)) {
+    struct edge_event evt = {0};
 
-void magazine_calculate_loss(void)
-{
-  const int last_evt_no = mag_state.evt_no ? 0 : 1;
-  const struct probe_window *current_event = &mag_state.probe_events[mag_state.evt_no];
-  const struct probe_window *last_event = &mag_state.probe_events[last_evt_no];
-
-  // Select which slop type we can detect based on whether or not lash
-  // is possible
-  int32_t * const property = magazine_lash_possible() ?
-    &mag_state.lash : &mag_state.spacing;
-
-  // Which edge we use as a reference point for our slop calculation
-  // depends on which direction we are moving in.
-  if (magazine_direction()) {
-    *property = abs(current_event->outboard - last_event->outboard);
-  } else {
-    *property = abs(current_event->inboard - last_event->inboard);
+    queue_dequeue(&edge_events, &evt);
+    report_sensor_edge((uint8_t)MAG_SENSOR, evt.state, evt.position);
   }
 }
 
-void magazine_record_edge(enum magazine_edge_type edge)
+static void magazine_edge_detector(const bool magazine_alignment_on)
 {
-  // Update our edge events based on which direction we are travelling
-  // in, and whether or not it is a rising or falling edge
-  struct probe_window *current_event = &mag_state.probe_events[mag_state.evt_no];
-  if (magazine_direction() ^ (E_MAGAZINE_EDGE_TYPE_RISING == edge)) {
-    current_event->inboard = sys.position[C_AXIS];
-  } else {
-    current_event->outboard = sys.position[C_AXIS];
-  }
-}
-
-void magazine_loss_detector(const uint8_t magazine_alignment_on)
-{
-  // When the probe is detected, copy the current carousel position
-  // into the probe position. If we were not previously on a flag,
-  // begin recording a magazine flag event, otherwise, end the event
-  // and calculate the loss
-  if (magazine_alignment_on && !mag_state.on_probe) {
-    mag_state.on_probe = true;
-    mag_state.probe_events[mag_state.evt_no].direction = magazine_direction();
-    magazine_record_edge(E_MAGAZINE_EDGE_TYPE_RISING);
-    mag_state.mag_events++;
-  } else if (mag_state.on_probe) {
-    mag_state.on_probe = false;
-    magazine_record_edge(E_MAGAZINE_EDGE_TYPE_FALLING);
-    mag_state.evt_no = mag_state.evt_no ? 0 : 1;
-
-    magazine_calculate_loss();
-    request_report(REQUEST_SLOP_REPORT, 0);
+  // We shouldn't send reports directly from here (as this is in an
+  // interrupt context).  Instead, we queue them up and will send them
+  // in the main program loop via a report request
+  if (magazine_alignment_on != mag_state.on_probe) {
+    struct edge_event evt = {0};
+    evt.state = magazine_alignment_on;
+    evt.position = sys.position[C_AXIS];
```

Comment:
I think this one is pretty specialized and will always be on this axis.  In the future, we'll have to think about generalizing.